### DISCUSSION TITLE
Use with React?

### DISCUSSION THREAD

**Discussion by keeper2597** [leolabs/ableton-js#70] [Q&A] [answered]
I'm trying to incorporate this into a React app, but I'm having trouble with the module resolution.  Keep getting errors that modules like gram, zlib, fs can't be resolved.  When I run a simple node app I don't get this problem. Any idea what I'm messing up?  Thanks for any suggestions.

`ERROR in ./node_modules/ableton-js/index.js 183:30-46
Module not found: Error: Can't resolve 'dgram' in '/Users/...`

**Comment by leolabs** [comment, marked as answer]
Hey @keeper2597,

Unfortunately, ableton-js doesn't work in the browser since it uses UDP to communicate with the Python plugin in Live. You'd need to create a server in Node that communicates with Live using ableton-js and offers an HTTP or WebSocket API that your React app can use. This is how I do it in AbleSet as well :)

**Reply by jbflow** [reply]
Hey @leolabs 

Really great project, not sure how I've not found this before. Just wondering why you're using UDP and not TCP. I explored web sockets in python in the remote scripts before but ran in to issues with it blocking so gave up, but that was when it was still on Python 2 back in Live 9 I think, and when I had far less experience in Python. Just wondering if it's possible to run a TCP server on a background thread to enable this to work in the browser for single page applications built in react etc.

Context: I'm building a react application that currently communicates with Live over MIDI (IAC ports), but this looks like a much better method. I would be happy to contribute to this project too as I have been coding in the Live Remote Script API for around 5 years and have extensive JavaScript and TypeScript knowledge.

**Comment by keeper2597** [comment]
Understood - I just found the issue raised a year ago that points to socketio as a good solution.  I'm looking in that direction now. 

Thanks for your great work!

**Comment by jbflow** [comment]
Hey @leolabs  & @keeper2597 

As mentioned in my above comment, I have managed to build a TCP Websocket server inside the Remote script and run it using threading.Thread, and I've sent some compressed data back and forth from a React application without needing to run a separate NodeJS Server. Seems to be a viable option at the moment and is working quite well. The code I've written is for my own project, but I will try and incorporate it into this project code and open a PR for review.

**Comment by jbflow** [comment]
@leolabs & @keeper2597 draft PR created [here](https://github.com/leolabs/ableton-js/pull/119)
